addons/website/models/res_users.py

```diff
@@ -95,3 +95,10 @@ def authenticate(cls, db, login, password, user_agent_env):
                 visitor_pre_authenticate_sudo.access_token = user_partner.id
                 visitor_pre_authenticate_sudo._update_visitor_last_visit()
         return uid
+
+    @api.constrains('groups_id')
+    def _check_one_user_type(self):
+        super()._check_one_user_type()
+        for user in self:
+            if user._is_internal() and user.website_id:
+                raise ValidationError(_("Remove website on related partner before they become internal user."))
```

Comment:
I am afraid of perfs here in case of huge `self` recordset. (i see that it's a concern in super() related methods)
There is multiple ways to avoid that: filtering users having `website_id` set before entering the loop, raw sql, single orm query search etc

I'd go with the "single orm query search" here, something like
```py
internal_users = self.env.ref('base.group_user').users & self
if any(user.website_id for user in internal_users):  # Note that this will shortcircuit once it face a website_id, which .filtered() wouldn't
    raise ValidationError(_("..."))
```
Maybe there is even better than this (without going into raw sql)